vpt/processors/speech_detector/__init__.py

```diff
@@ -1,6 +1,83 @@
 '''Processes an audio stream to detect speech.'''
+
+import numpy as np
+from rx.subject import Subject
+from scipy import fft
+
+from vpt.sources.base import SourceBase
 from vpt.processors.base import ProcessorBase
 
+# The frequency range of human voice in Hz.
+VOICE_RANGE = range(80, 1100 + 1)
+# The threshold which the amplitude needs to overcome to be considered present (in undefined units).
+NOISE_THRESHOLD = .5
+SILENCE_THRESHOLD = .3
 
-class SpeechDetector(ProcessorBase):
+
+class SpeechDetector(ProcessorBase[bool]):
     '''Detects speech in an audio stream and outputs a signal stream for that.'''
+    subj = Subject()
+    mean_loudness: np.float64
+    _samples: int
+
+    def __init__(self, audio_source: SourceBase[np.ndarray]):
+        '''Wire up the detector to an arbitrary audio source.'''
+        audio_source.get_data_stream().subscribe(self.detect_speech)
+        self.mean_loudness = None
+        self._samples = 0
+
+    def get_data_stream(self):
+        return self.subj
+
+    def detect_speech(self, frame: np.ndarray):
+        '''Detect sound in speech frequencies in a given audio frame.'''
+        if frame.ndim == 1:
+            frame = frame.reshape(len(frame), 1)
+
+        freqs = self.fft(frame.mean(axis=1))
+        this_loudness = freqs.max(axis=0)
+
+        if self.is_silence(this_loudness):
+            self.subj.on_next(False)
+        else:
+            freqs_norm = self.normalize_frequencies(freqs)
```

Comment:
`normalize_frequencies` is a static method. Call it from the class rather than the instance

Reply:
Same for `fft`

Reply:
Why? It requires hardcoding the name of the class that may change, while `self` is not bound to anything

Reply:
So what if it changes? The IDE (or pylint) will show an error, so one cannot forget to change it. It's not a big deal, but it just feels more correct.
Also, it matters in other cases where it is not a method but some other field that can be overwritten on the instance but not the class (or vice versa) and then it causes confusion

Reply:
What if I work in a text editor with no linter? No, relying on IDE tools to make code decisions is a bad idea. It feels more correct, I agree, but this feels more robust. I hate to hardcode stuff like that.
About that second thing – if that is the case, that in itself is a major red flag that the code is messed up

Reply:
If you work in a text editor with no linter then install one. smh

How often do we rename classes? Don't you think that the code that will rely on this class will have to be edited also? Then your argument that you'll need not edit several lines less doesn't really hold, because you'll have to change code anyway. Readability > a little less lines to edit if ever needed.

Reply:
I don't really see how this adds to readability, and the benefit of maintainability is evident

Reply:
> Don't you think that the code that will rely on this class will have to be edited also?

The difference is that it is expected that if you rename a class, you go over places where you use this class and change the name. It is unexpected that the class itself relies on its name, that's a very strange and fragile binding

Reply:
Listen, you don't edit python code by hand. That's ridiculous. You edit python code via tools. And there's a tool that allows you to rename a class. And it will work in both cases. But when you write the static method with the name of the class, you make it _evident_ that this method is static and therefore presents some generic algorithm related to the class, not to a particular instance. That is, if this method is static then the current instance cannot be modified, and this **is** a valuable information to the reader.  So, with the change we have better readability of the code, which makes the mentioned maintenance easier also.

Reply:
Yes, there are benefits to calling static methods from classes in terms of code readability, and there are downsides. FYI my editor doesn't have a tool to rename a class, and I would never use it anyway.

I don't want to change this code to call from the class instead of the instance. Maybe this will settle the discussion.